baseband/guppi/tests/test_guppi.py

```diff
@@ -0,0 +1,544 @@
+        with guppi.open(filename, 'ws', time=h.time, bps=h.bps,
+                        sample_rate=h.sample_rate,
+                        pktsize=h['PKTSIZE'], overlap=0,
+                        payload_nbytes=self.header_w.payload_nbytes // 4,
+                        nchan=2, npol=1) as fw:
+            fw.write(self.payload[:self.spf, 0, :2])
+            assert np.abs(fw.start_time - start_time) < 1.*u.ns
+            assert (np.abs(fw.time - (start_time + self.spf / (2500 * u.Hz))) <
+                    1. * u.ns)
+
+        with guppi.open(filename, 'rs') as fh:
+            data_onepol = fh.read()
+            assert np.abs(fh.start_time - start_time) < 1.*u.ns
+            assert np.abs(fh.stop_time - (start_time + self.spf /
+                                          (2500 * u.Hz))) < 1.*u.ns
+        assert np.all(data_onepol == self.payload[:self.spf, 0, :2])
+
+        # Try reading a single polarization.
+        with guppi.open(SAMPLE_FILE, 'rs', subset=0) as fh:
+            assert fh.sample_shape == (4,)
+            assert fh.subset == (0,)
+            record3 = fh.read(12)
+            assert np.all(record3 == record[:12, 0, :4])
+
+        # Read right polarization, but read channels in reverse order.
+        with guppi.open(SAMPLE_FILE, 'rs', subset=(1, [1, 0])) as fh:
+            assert fh.sample_shape == (2,)
+            data_sub = fh.read()
+            assert np.all(data_sub[:, 1] == record[:, 1, 0])
+
+        # Test that squeeze attribute works on read (including in-place read).
+        with guppi.open(filename, 'rs', squeeze=False) as fh:
+            assert fh.sample_shape == (1, 2)
+            assert fh.sample_shape.npol == 1
+            assert fh.sample_shape.nchan == 2
+            assert fh.read(1).shape == (1, 1, 2)
+            assert fh.read(10).shape == (10, 1, 2)
+            fh.seek(0)
+            out = np.zeros((12, 1, 2), dtype=np.complex64)
+            fh.read(out=out)
+            assert fh.tell() == 12
+            assert np.all(out.squeeze() == self.payload[:12, 0, :2])
+
+        # Test that squeeze=False works on write (we checked above it works
+        # with squeeze=True).
+        with guppi.open(filename, 'ws', time=h.time, bps=h.bps,
+                        sample_rate=h.sample_rate,
+                        pktsize=h['PKTSIZE'], overlap=0,
+                        payload_nbytes=self.header_w.payload_nbytes // 4,
+                        nchan=2, npol=1, squeeze=False) as fw:
+            fw.write(self.payload[:self.spf, 0:1, :2])
+            assert np.abs(fw.start_time - start_time) < 1.*u.ns
+            assert (np.abs(fw.time - (start_time + self.spf / (2500 * u.Hz))) <
+                    1. * u.ns)
+
+        with guppi.open(filename, 'rs', squeeze=False) as fh:
+            data_onepol = fh.read()
+        assert np.all(data_onepol.squeeze() == self.payload[:self.spf, 0, :2])
+
+    # Test that writing an incomplete stream is possible, and that frame set is
```

Comment:
How can this work? I thought GUPPI has no way to tell whether a frame is valid or not. I think this ends up just testing that if you write zeros, you get them back...

Reply:
It does, and the comment is wrong.  I've corrected it, and will correct DADA as well in the other PR.